Bazel fixes

asa opened the pull request:
These are a few fixes I needed to get the included bazel file to play nice with OpenImageIO.
I believe they are general enough to be included for others.

kdt3rd committed:
Add initial context objects

These are not final, but should represent concepts we will use in new
world of using the core to replace the internals of the C++

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Initial conversion of input file to use core

This is not complete (some of the dependencies are circular, so depends
on other later conversions to finish), so represents a WIP state

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Complete conversion of header attributes to header class

Add additional constructor to opaque attribute to enable this.
Registered custom attributes still TODO

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Better handle streams and attempts to read past end

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
handle custom attributes

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
switch to c++17 by default

This will enable use of string_view and similar API convenience as makes
sense. Given 17 has been standard for 4 years of VFX Platform, this
should be ok.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
write some initial doxygen documentation

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove tracking of current position

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Improve exception messages

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add test, accessor to match multipart behavior

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Clean extraneous deep data during chunk reconstruction

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
fix checkfile typo, return value for ripmap vs mipmap tiles

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Convert MultiPartInputFile to use context

Also simplifies InputFile to always defer to MultiPartInputFile to start
Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove extra spaces

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Reproduce check against the count of chunk items

The C++ library had arbitrary (duplicate, non-editable) checks against a
max chunk count, because the IO streams can't always provide the file
size as a safety check. reproduce that here

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Address review notes from meshula

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Small cleanup

reduce scope of variables, collapse logic to avoid duplicate conditional
blocks, re-enable failable free test (original issues had already been
fixed, but test had not been re-enabled)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove commented deprecations until further review

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
add try / catch to avoid throwing through the C interface

Streams that throw are an unfortunate choice, but such is the way.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
make b44 consistent with compression rules

Not sure why this was coded this way, must have been a separate /
previous bug, but b44 / b44a are not exempt from the rule that if the
compression is larger, store uncompressed.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
make y and x sampling use consistent functions

Avoid duplicate code, simplify logic, fix scenario for all negative y
coordinates

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add channel lookup utility

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Allow the unpack routines to splat into a subset of the chunk

This will avoid an extra temporary buffer to be cached when reading
subsets of the image chunks

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Convert scanline input file to use the core

simplifies input file as well. Future changes may lift common code out
of scanline input into common utility in context / somewhere.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Actually compare two images

test code was comparing one image to itself, fix this, but the ratio
check was too narrow for half float values

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix scope of threading enable check

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Make it easier for c++ to report specific exceptions on missing

When a chunk is missing, c++ would test if the offset is 0 and throw a
different exception, replicate that here

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add query to get the tile counts for a level

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Greatly simplify the cache buffer mechanism when input is tiled

This also adds a bypass to avoid double copy, simpler memory management,
and prepping to have some common code refactored

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove sparsely used cache variable

No need to cache a variable only used in an exceptional case, start to
prep for refactor of data block into context

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Convert tiled input file to use the core library

Once all the input classes are done, will refactor to lift code out of
the internals of data to common elements in a singular place to avoid
code duplication.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add ability for context initializer to tiled rgba input file

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
avoid potential (not likely) overflow in tile count calc

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add a temporary context

This will allow us to have a context without having to actually have a
file (i.e. to use a header object in the C++ layer, or compression)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Expose compression / decompression pipeline function

This will allow users of the pipeline to run the compression /
decompression, and to replace the C++ compressor utilities. This also
adds some additional checks and looser memory restrictions to the
decompression scheme for ease when we don't have all the information
needed for encoding / decoding.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add a tag to the constructors to indicate read / write / temp

This makes construction consistent with core, where by once constructed,
the object is ready to be used. (file is opened and parsed or ready for
population)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
switch compression to use core encode / decode pipeline function

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Ensure common type for std::max

Not all compilers coerce the types to the same

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove unneeded vararg from macro

not all compilers have named vararg macros yet

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove friend cheat as windows symbols change

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
avoid potential UB math, ensure value is initialized

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Avoid oss fuzz checks in c++ by initializing scans differently

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove over-optimization

Previous commits had tried to reduce the number of calls to chunk offset
but that prevented updates during partial header changes

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Move tables, test logic to core now compression is moved

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix funky formatting

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Invert stat and open to do open and use fstat

This should fix 2 of the code scan security warnings, at no particular
cognitive overhead, although is just in tests anyway

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Properly fail when version number is not 1

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix the sample stride and set the pointer to null instead of alloc size 0

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Easier to find bugs with null pointer instead of size 0 alloc

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Reduce scope of messages and avoid set but unused when threading off

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Convert deep scanline input file to use core library

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Relax safety check when reading uncompressed deep

tiled deep files which are uncompressed incorrectly write a "full" tile
worth of sample counts, except the data is actually tight packed to the
smaller portion of that tile size (i.e. at coarser levels or at
boundaries of non-tile multiple images). so handle that check against
the memory size when copying uncompressed data to unpack (i.e. when in
the non-cumulative count mode)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove extraneous mutex

mutex is now has-a, not is-a but the subclass hadn't been removed

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add consistent checks, fix messaging

- readPixels and readPixelSampleCounts were not consistent in frame
buffer checks
- pixelData is allowed to be null but wasn't being guarded against
- Clear up message around check being run to apply the decode chunk
cache
- add common optimization for sample count copying

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Convert deep tiled input to use the core library

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix compile when threading disabled

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove no-longer needed input part data

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove now-unneeded legacy stream adapter

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
fix thread race condition

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
fix sanitizer warning

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
deep is not doing any optimization checks on the user bytes, let them vary as a result

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
minor optimization

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
pass the sample stride through instead of just the bytes like normal image reading

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
msvc does not really have stdatomic support yet, work around same as elsewhere

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
better handle under/overflow of chunk table in deep tiles

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

asa committed:
adding the includes path above the sub components allows bazel to import items with <Iex/foo.h> include paths

Signed-off-by: asa <asa@automaton.is>

asa committed:
the OpenEXR_VERSION_PATCH should be renamed to all caps OPENEXR_VERSION_PATCH otherwise the bazel template renames don't work

Signed-off-by: asa <asa@automaton.is>

asa committed:
favor Imath

Signed-off-by: asa <asa@automaton.is>

Vertexwahn commented on BUILD.bazel:
```diff
@@ -501,7 +506,7 @@ cc_library(
     deps = [
         ":IlmThread",
         ":OpenEXRCore",
-        "@imath//:Imath",
```
Module names are typically written in lowercase. See for example [here](https://github.com/bazelbuild/bazel-central-registry/tree/main/modules) - no module is written in uppercase. I think this is done to prevent OS problems - for instance Windows vs. Linux - Windows would maybe treat imath, Imath, IMATH the same way. Change it back to` "@imath//:Imath",`

asa replied:
I agree, except that openexr uses a reference to the imath bazel module dep that capitalizes it.

asa replied:
In the bazel central registry that is.

asa replied:
eg: https://github.com/bazelbuild/bazel-central-registry/blob/main/modules/openexr/3.2.4/MODULE.bazel#L11

Vertexwahn commented:
@asa imath@3.1.12 is just landing in the BCR: https://github.com/bazelbuild/bazel-central-registry/pull/2772

kdt3rd committed:
legacy library allowed empty part name

Also add unique name validation

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
tell python wheels to use c++17

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
only allow cxx standard to override if greater than 17

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Tell wheel builds to use new-enough mac O.S. to deal with c++17

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
set output bit in make_ofstream

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
simplify window filestream creation

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Allow other things to share access to this file under windows

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove now unreachable code

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
improve error reporting in stream wrapper

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix compile warnings

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

dependabot[bot] committed:
Bump pypa/cibuildwheel from 2.20.0 to 2.21.0 (#1819)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.20.0 to 2.21.0.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/bd033a44476646b606efccdd5eed92d5ea1d77ad...79b0dd328794e1180a7268444d46cdf12e1abd01)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github/codeql-action from 3.26.6 to 3.26.7 (#1820)

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 3.26.6 to 3.26.7.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/4dd16135b69a43b6c8efb853346f8437d92d3c93...8214744c546c1e5c8f03dde8fab3a7353211988d)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kdt3rd committed:
Address some initial fuzz reports after core rewrite merge (#1821)

* Add validation that the types of the attributes are correct, not just that they exist

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* use constructor semantics to zero init the context pointer, avoid one memory sanitizer warning

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* to avoid UB with streams, keep attr sizes with int32 size

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* handle an empty buffer cleanly

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

---------

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add duplicate validation

There is validation for tile descriptions already when looking at a
tiled file, but if someone provides a tile description in a non-tiled
file, validate that the enums will be valid

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
avoid calling compression when there is nothing to do

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
avoid decoding when there is an empty chunk

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Change how when channels are validated, duplicated

Old library allowed names to accumulate, whether that is a great feature
or not, preserve that when not in strict mode.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

fnordware committed:
Add includes for Windows (#1827)

Signed-off-by: Brendan Bolles <brendan@fnord.com>

dependabot[bot] committed:
Bump github/codeql-action from 3.26.7 to 3.26.8 (#1833)

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 3.26.7 to 3.26.8.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/8214744c546c1e5c8f03dde8fab3a7353211988d...294a9d92911152fe08befb9ec03e240add280cb3)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

pedroborgescruz committed:
Fixed formatting issue in website/OpenEXRFileLayout.rst (#1832)

* Update OpenEXRFileLayout.rst

Signed-off-by: Pedro Borges <pedroborgespc@gmail.com>

* Update OpenEXRFileLayout.rst

Signed-off-by: Pedro Borges <pedroborgespc@gmail.com>

* Update OpenEXRFileLayout.rst

Signed-off-by: Pedro Borges <pedroborgespc@gmail.com>

---------

Signed-off-by: Pedro Borges <pedroborgespc@gmail.com>

cary-ilm committed:
Update sigstore release signing action (#1822)

The default behavior of sigstore/gh-action-sigstore-python has
changed. Disable the automatic uploading of signed artifacts, since
this now includes artifacts named with just the tag, without the
"openexr-" prefix.

Also, the signature file now has a .json suffix.

Signed-off-by: Cary Phillips <cary@ilm.com>

dependabot[bot] committed:
Bump pypa/cibuildwheel from 2.21.0 to 2.21.1 (#1824)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.21.0 to 2.21.1.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/79b0dd328794e1180a7268444d46cdf12e1abd01...d4a2945fcc8d13f20a1b99d461b8e844d5fc6e23)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cary-ilm commented:
Where do we stand with this? The CI build check is failing with:

`ERROR: no such package '@@[unknown repo 'Imath' requested from @@]//': The repository '@@[unknown repo 'Imath' requested from @@]' could not be resolved: No repository visible as '@Imath' from main repository`

A new release is coming next week, so it would be good to get this sorted out soon. Thanks!

asa commented:
The issue is the BCR is remapping `imath -> Imath` [here](https://github.com/bazelbuild/bazel-central-registry/blob/faca465caf067a601f0a6f47fee6fabb607ba6d4/modules/openexr/3.2.4/MODULE.bazel#L11) at the module namespace level (which is trickling down into the include namespace so that include `<Imath/foo.h>` works).

Paths forward are either 

1. update the BCR stanza  to use lowercase imath, not to redirect, and also update openexr deps to "@imath", and include paths to use lowercase `<imath/foo.h>`. This would allow both people using bazel modules and non module setups to work.

2. keep the current setup where the BCR is remapping `imath -> Imath`, and do some work on the openexr BUILD files to support both variants of  include prefixes `<Imath <imath`, so that all the platforms work.

asa commented:
I took a stab at the BCR changes needed [here:](https://github.com/bazelbuild/bazel-central-registry/pull/2822).
Running into some cache staleness weirdness with the overlay setup... not sure what I am doing wrong there...

asa committed:
lower case

Signed-off-by: asa <asa@automaton.build>

cary-ilm commented:
Looks like that may have fixed it, thanks!

Our project policy is to require both a signed Contributor License Agreement and signed commits, a.k.a "Digital Certificate of Origin". Click on the red NOT COVERED button in the EasyCLA box above to sign the legal agreement. If you're doing the work for your employer, or your company has a policy around open source contributions, a legal representative of the company will need to sign the corporate version of the document.

For the DCO, commits need to be created with the `-s` flag, which added a "signed off by" line with your name and email. You can add this to the existing commits with `git commit --amend -s`. But you'll need to do this for each commit so it may be easier to squash the 5 commits into one.  Once you get the commits signed, you'll need to do a `git push --force` to push them to the PR's branch.

Apologies for the requirements, but is Academy Software Foundation policy. Thanks!

asa commented:
closing in favor of (#1836)
